### ISSUE TITLE
Types Compatibility - Foreach source to iterate over

### ISSUE THREAD

**Issue by craiglondon** [kalessil/phpinspectionsea#827]
This is not really an issue, or maybe its a false positive.

How do I write the `@var` syntax for a value that is deeply nested?

The message the inspection is reporting is `Expressions' type was not recognized, please check type hints.`

Here is the code
`foreach ($this->widgetObject->widgetArray['subArray'] as $subArray) {`

$this = Object
widgetObject = Object
widgetArray = Array
subArray = Array

`@var $this array[][][] ` definitely doesn't work here.

Maybe it isn't possible to write a `@var` for this deeply nested value?

**Comment by kalessil** [comment]
@craiglondon : the right way is to annotate the both widgetObject and widgetArray with relevant type information.

If it's not possible, then just introduce a local variable:
```php
        /* @var array|array[] $widgetObject*/
        $array = ...;
        foreach ($array as $subArray) {}
```
or, there is less atractive option (but it can chabge semantics if source is traversable, not an array):
```php
        foreach ((array)... as $subArray) {}
```

**Comment by craiglondon** [comment]
@kalessil Thank you for the response.  I will apply the information you provided.

**Comment by kalessil** [comment]
You are welcome =) Which of options did you like more (asking as probably review our docs for the inspection as follow up on this ticket).

**Comment by craiglondon** [comment]
Maybe this documentation could be enhanced?

https://github.com/kalessil/phpinspectionsea/blob/master/docs/types-compatibility.md#foreach-source-to-iterate-over

I still am not able to always write `@var` declarations that will satisfy that rule.

`/* @var array[] $this->type['selectOptions'] */`
`foreach ($this->type['selectOptions'] as $option) {`

$this = object (this Model)
type = array
selectOptions = array

**Comment by craiglondon** [comment]
`foreach ((array)... as $subArray) {}` . This works, but I don't like the potential side effect of casting the variable if it isn't an array.

**Comment by kalessil** [comment]
I'll enhance the docs as well. 

Also: in my snippet and docs as well you have to introduce a **local variable** and annotate its type.
Annotating complex expressions is not supported by PhpStorm (and perhaps even not part of PhpDoc).

Would you please give a try to a local variable?

**Comment by kalessil** [comment]
Re-opening the issue as it was closed not intentionally.

**Comment by craiglondon** [comment]
@kalessil I will try... would this be the correct syntax for this example below?

```
/* @var array $option */
foreach ($this->type['selectOptions'] as $option) {
```

**Comment by craiglondon** [comment]
re: `Annotating complex expressions is not supported by PhpStorm (and perhaps even not part of PhpDoc).`  I concur, I don't think the complex expressions are supported by PHPDoc or PhpStorm.

**Comment by kalessil** [comment]
@craiglondon : correct one would be

```php
/* @var array[] $options */
$options = $this->type['selectOptions'];
foreach ($options as $option) {
    ...
}
```

**Comment by kalessil** [comment]
Done: docs has been extended with the type annotation example.

**Comment by craiglondon** [comment]
@kalessil Appreciate it!  Thank you.

**Comment by kalessil** [comment]
@craiglondon : you are welcome =) In case if you run in something similar (what needs documenting), just file an issue.
